amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+            --device {{platform._part}}
+            --family {{platform.device}}
+            --json {{name}}.json
+            --freq 27
+            --cst {{name}}.cst
+            --write {{name}}.json
+        """,
+        r"""
+        {{invoke_tool("gowin_pack")}}
+            {{verbose("-v")}}
+            -d {{platform.device}}
+            -o {{name}}.fs
+            {{get_override("gowin_pack_opts")|options}}
+            {{name}}.json
+        """
+    ]
+
+    ### Gowin vendor toolchain templates
+
+    _gowin_required_tools = ["gw_sh"]
+
+    @property
+    def _device_name(self):
+        return "{}-{}{}".format(self.series, self.size, self.subseries)
+
+    _gowin_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        **_common_file_templates,
+        "{{name}}.tcl": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v",".sv",".vhd",".vhdl") -%}
+                add_file {{file}}
+            {% endfor %}
+            add_file -type verilog {{name}}.v
+            add_file -type cst {{name}}.cst
+            add_file -type sdc {{name}}.sdc
+            set_device -name {{platform._device_name}} {{platform._part}}
+            set_option -verilog_std v2001 -print_all_synthesis_warning 1 -show_all_warn 1
+            {{get_override("add_options")|default("# (add_options placeholder)")}}
+            run all
+        """,
+    }
+
+    _gowin_command_templates = [
+        r"""
+        {{invoke_tool("gw_sh")}}
+            {{name}}.tcl
+        """,
+        "cp impl/pnr/project.fs {{name}}.fs"
+    ]
+
+    def __init__(self, *, toolchain="Apicula"):
+        super().__init__()
+
+        assert toolchain in ("Apicula", "Gowin")
+        self.toolchain = toolchain
+
+    @property
+    def _toolchain_env_var(self):
+        if self.toolchain == "Apicula" or self.toolchain == "Gowin":
+            return f"AMARANTH_ENV_{self.toolchain}"
+        assert False
+
+    @property
+    def required_tools(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_required_tools
+        elif self.toolchain == "Gowin":
+            return self._gowin_required_tools
+        assert False
+
+    @property
+    def file_templates(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_file_templates
+        elif self.toolchain == "Gowin":
+            return self._gowin_file_templates
+        assert False
+
+    @property
+    def command_templates(self):
+        if self.toolchain == "Apicula":
+            return self._apicula_command_templates
+        elif self.toolchain == "Gowin":
+            return self._gowin_command_templates
+        assert False
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "true"
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "sys_clk0":
```

Comment:
Why `sys_clk0`?

Reply:
Not sure anymore, I think I got that from the quicklogic / quickfeather bits of amaranth-lang and amaranth-boards. Do you have a better suggestion?

Reply:
`OSC` or `OSCx`

Reply:
Ok, fixed.